File: apps/ios/App/Import/ImportScreen.swift
import SwiftUI
import TrackEvolutionKit
import UniformTypeIdentifiers

/// Pick a video on the phone and get lap times out of it (NS-30).
///
/// The web app's **Import video / telemetry…** does this on a laptop; this does it
/// on the device that already has the footage — a GoPro clip arrives over Wi-Fi
/// into Photos, a Corvette PDR clip lands in Files off a USB stick, and both are
/// usually there before the laptop is opened.
///
/// Two phases on one screen: choose, then review. The review half is the *same*
/// `ReviewScreen` a stopped recording goes through, because the laps, the line
/// picker and the save are the same job.
struct ImportScreen: View {
    /// The event this import came from, pre-selected in the review's event picker.
    var eventId: Int?
    /// A clip handed straight to the app by Files or the share sheet.
    var incoming: URL?

    @Environment(AppRouter.self) private var router
    @Environment(AuthController.self) private var auth
    @State private var model = ImportModel()
    @State private var showingFileImporter = false
    @State private var showingPhotoPicker = false
    @State private var showingPaywall = false

    /// The Pro gate (NS-32 rule 5), from the cached entitlement, so an import in a
    /// paddock with no signal proceeds for a driver who was Pro at the last sync.
    /// Off until phase D flips `Entitlement.gatesEnabled`.
    private var gate: ProGate.Decision {
        ProGate.decide(.videoImport, entitlement: auth.entitlement)
    }

    var body: some View {
        Group {
            if let clips = model.clips {
                ReviewScreen(source: .imported(clips), preferredEventId: eventId, onFinish: { router.popToRoot() })
            } else {
                ScrollView {
                    if gate == .paywall {
                        gated
                            .padding(TESpacing.pageGutter)
                    } else {
                        chooser
                            .padding(TESpacing.pageGutter)
                    }
                }
                .background(Color(.bgPage))
                .navigationTitle("Import video")
                .navigationBarTitleDisplayMode(.inline)
            }
        }
        // Keyed on the gate: a clip handed over by Files while free is parsed the
        // moment a subscription made from the sheet lands, not lost.
        .task(id: gate) {
            if gate == .proceed, let incoming, model.clips == nil, !model.isParsing {
                await model.parse([PickedVideo(name: incoming.lastPathComponent, url: incoming)])
            }
        }
    }

    /// What a free account sees instead of the pickers: the paywall, one tap away,
    /// rather than disabled buttons that explain nothing.
    private var gated: some View {
        VStack(alignment: .leading, spacing: TESpacing.gridGap) {
            TECard {
                VStack(alignment: .leading, spacing: 10) {
                    Text("Telemetry import is a Pro feature")
                        .teStyle(.h3)
                        .foregroundStyle(Color(.textStrong))
                    Text("""
                        Corvette PDR and GoPro clips carry telemetry alongside the picture. Track \
                        Evolution Pro reads it on this phone — the video never leaves it — and turns \
                        it into lap times and channel graphs.
                        """)
                        .teStyle(.sm)
                        .foregroundStyle(Color(.textMuted))
                    Button("See Track Evolution Pro") { showingPaywall = true }
                        .buttonStyle(TEButtonStyle(kind: .accent))
                        .accessibilityIdentifier("importPaywall")
                        .sheet(isPresented: $showingPaywall) {
                            PaywallSheet(context: .videoImport)
                        }
                }
            }
        }
    }

    @ViewBuilder
    private var chooser: some View {
        VStack(alignment: .leading, spacing: TESpacing.gridGap) {
            TECard {
                VStack(alignment: .leading, spacing: 10) {
                    Text("Lap times from video")
                        .teStyle(.h3)
                        .foregroundStyle(Color(.textStrong))
                    Text("""
                        Corvette PDR and GoPro clips carry telemetry alongside the picture. \
                        Pick one and the laps come out of it here — the video never leaves \
                        this phone and is never copied; only its telemetry track is read.
                        """)
                        .teStyle(.sm)
                        .foregroundStyle(Color(.textMuted))

                    if model.isParsing {
                        HStack(spacing: 10) {
                            ProgressView()
                            Text(model.progress)
                                .teStyle(.sm)
                                .foregroundStyle(Color(.textMuted))
                        }
                        Button("Cancel") { model.cancel() }
                            .buttonStyle(TEButtonStyle(kind: .quiet))
                    } else {
                        filesButton
                        photosButton
                    }

                    if let failure = model.failure {
                        TEErrorBanner(message: failure)
                    }
                }
            }

            TECard {
                VStack(alignment: .leading, spacing: 6) {
                    Text("What works")
                        .teStyle(.eyebrow)
                        .foregroundStyle(Color(.textMuted))
                    Text("""
                        A PDR clip recorded with the track's beacon arrives with exact lap \
                        times and needs nothing from you. A GoPro clip, or a PDR one without \
                        beacons, has GPS but no lap markers — tap where the start/finish line \
                        is and every pass across it is timed.
                        """)
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                    Text("""
                        `.vbo` and other logger files stay on the web app, where the screen is \
                        bigger and the SD card is already in the laptop.
                        """)
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textFaint))
                }
            }
        }
    }
    /// Each picker lives on its own view with exactly **one** presentation
    /// modifier, and only in the chooser phase.
    ///
    /// Two modals attached to one view is a documented way to wedge SwiftUI here
    /// (see `apps/ios/README.md`), and the obvious shape — `.fileImporter` and
    /// `.sheet` both hung off this screen's root — would have kept them attached
    /// through the review phase too, next to the review's own confirmation dialog.
    /// Splitting them costs nothing and keeps the rule intact.
    private var filesButton: some View {
        Button("Choose from Files") { showingFileImporter = true }
            .buttonStyle(TEButtonStyle(kind: .accent))
            .accessibilityIdentifier("importFromFiles")
            .fileImporter(
                isPresented: $showingFileImporter,
                // Multiple selection from the start: a session is often several
                // clips, and the batch is what lets a beacon-timed PDR recording
                // re-anchor a beacon-less one beside it.
                allowedContentTypes: [.mpeg4Movie, .quickTimeMovie, .movie],
                allowsMultipleSelection: true
            ) { result in
                switch result {
                case .success(let urls):
                    Task {
                        await model.parse(urls.map { PickedVideo(name: $0.lastPathComponent, url: $0) })
                    }
                case .failure(let error):
                    model.failure = error.localizedDescription
                }
            }
    }

    private var photosButton: some View {
        Button("Choose from Photos") {
            Task {
                if await model.authorizePhotos() { showingPhotoPicker = true }
            }
        }
        .buttonStyle(TEButtonStyle(kind: .quiet))
        .accessibilityIdentifier("importFromPhotos")
        .sheet(isPresented: $showingPhotoPicker) {
            PhotoVideoPicker(
                onPicked: { ids in
                    showingPhotoPicker = false
                    Task { await model.parsePhotos(ids) }
                },
                onCancel: { showingPhotoPicker = false }
            )
            .ignoresSafeArea()
        }
    }
}

/// Picking and parsing, and the states in between.
///
/// The parse runs in one task off the main actor with the `FileHandle` confined to
/// it, and checks for cancellation between samples — a driver who picked the wrong
/// 4 GB clip must be able to back out.
@MainActor
@Observable
final class ImportModel {
    /// nil until a selection has been parsed; then the review takes over.
    private(set) var clips: [ImportedClip]?
    private(set) var isParsing = false
    private(set) var progress = ""
    var failure: String?

    private var task: Task<Void, Never>?

    func authorizePhotos() async -> Bool {
        if await PhotoLibraryVideos.authorize() { return true }
        failure = PickerFailure.photoAccessDenied.message
        return false
    }

    /// Resolve picked assets to files without downloading anything, then parse.
    func parsePhotos(_ localIdentifiers: [String]) async {
        var picks: [PickedVideo] = []
        var problems: [String] = []
        for id in localIdentifiers {
            do {
                picks.append(try await PhotoLibraryVideos.fileURL(for: id))
            } catch let failure as PickerFailure {
                problems.append(failure.message)
            } catch {
                problems.append(error.localizedDescription)
            }
        }
        failure = problems.isEmpty ? nil : problems.joined(separator: "\n\n")
        if picks.isEmpty { return }
        await parse(picks)
    }

    func parse(_ picks: [PickedVideo]) async {
        guard !picks.isEmpty else { return }
        task?.cancel()
        isParsing = true
        progress = "Reading telemetry from \(picks.count) file\(picks.count == 1 ? "" : "s")…"

        let work = Task { [picks] in
            do {
                let parsed = try await TelemetryImporter.parse(picks)
                await MainActor.run {
                    self.clips = parsed
                    self.isParsing = false
                }
            } catch is CancellationError {
                await MainActor.run { self.isParsing = false }
            } catch {
                await MainActor.run {
                    self.failure = error.localizedDescription
                    self.isParsing = false
                }
            }
        }
        task = work
        await work.value
    }

    func cancel() {
        task?.cancel()
        task = nil
        isParsing = false
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/videoimport/ImportScreen.kt
package app.trackevolution.videoimport

import android.content.ContentResolver
import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import app.trackevolution.ui.TEErrorBanner
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

/**
 * Pick a video on the phone and get lap times out of it.
 *
 * The web app's **Import video / telemetry…** does this on a laptop; this does
 * it on the device that already has the footage — a GoPro clip lands in the
 * camera roll over Wi-Fi, a Corvette PDR clip comes off the USB stick into
 * Files, and both are usually there before the laptop is opened. iOS got this
 * first (NS-30); the parsers are the same port, checked against the same
 * fixture, and the review that follows is the *same* screen a stopped
 * recording goes through, because the laps, the line picker and the save are
 * the same job.
 *
 * Only the choosing lives here. Once the clips are parsed [onParsed] hands them
 * to the review overlay and this destination is popped.
 */
@Composable
fun ImportScreen(
    model: ImportModel,
    /** Clips handed straight to the app by a share sheet, parsed on arrival. */
    incoming: List<Uri>?,
    onConsumedIncoming: () -> Unit,
    onParsed: (List<ImportedClip>) -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = TrackTheme.colors
    val type = TrackTheme.typography

    // The system document picker: Files, Downloads, the USB stick, and every
    // photo app that publishes a documents provider — Google Photos included.
    // `content://` URIs from here open as seekable descriptors, so nothing is
    // copied. Multiple selection from the start: a session is often several
    // clips, and the batch is what lets a beacon-timed PDR recording re-anchor
    // a beacon-less one beside it.
    val documents = rememberLauncherForActivityResult(ActivityResultContracts.OpenMultipleDocuments()) { uris ->
        if (uris.isNotEmpty()) model.parse(uris)
    }
    // The photo picker, for the camera roll on phones where the document picker
    // doesn't surface it. Same URIs, same no-copy read.
    val photos = rememberLauncherForActivityResult(ActivityResultContracts.PickMultipleVisualMedia()) { uris ->
        if (uris.isNotEmpty()) model.parse(uris)
    }

    LaunchedEffect(incoming) {
        val uris = incoming ?: return@LaunchedEffect
        onConsumedIncoming()
        if (uris.isNotEmpty()) model.parse(uris)
    }

    val clips = model.clips
    LaunchedEffect(clips) {
        if (clips != null) {
            model.consumeClips()
            onParsed(clips)
        }
    }

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(colors.bgPage)
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text("Import video", style = type.h1, color = colors.textStrong)

        TrackCard(Modifier.fillMaxWidth()) {
            Text("Lap times from video", style = type.h3, color = colors.textStrong)
            Text(
                "Corvette PDR and GoPro clips carry telemetry alongside the picture. Pick one and the " +
                    "laps come out of it here — the video never leaves this phone and is never copied; " +
                    "only its telemetry track is read.",
                style = type.sm,
                color = colors.textMuted,
                modifier = Modifier.padding(top = 6.dp, bottom = 12.dp),
            )

            if (model.isParsing) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    CircularProgressIndicator(modifier = Modifier.height(18.dp), color = colors.accent)
                    Text(
                        model.progress,
                        style = type.sm,
                        color = colors.textMuted,
                        modifier = Modifier.padding(start = 10.dp),
                    )
                }
                TextButton(onClick = model::cancel) {
                    Text("Cancel", style = type.sm, color = colors.textMuted)
                }
            } else {
                Button(
                    onClick = { documents.launch(arrayOf("video/mp4", "video/quicktime", "video/*")) },
                    modifier = Modifier.fillMaxWidth().testTag("importChooseVideos"),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = colors.accent,
                        contentColor = colors.accentContrast,
                    ),
                ) {
                    Text("Choose videos", style = type.bodyStrong)
                }
                OutlinedButton(
                    onClick = {
                        photos.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.VideoOnly))
                    },
                    modifier = Modifier.fillMaxWidth().padding(top = 6.dp).testTag("importChoosePhotos"),
                ) {
                    Text("Choose from Photos", style = type.bodyStrong, color = colors.textStrong)
                }
            }

            model.failure?.let { TEErrorBanner(it, modifier = Modifier.padding(top = 10.dp)) }
        }

        TrackCard(Modifier.fillMaxWidth()) {
            Text("WHAT WORKS", style = type.eyebrow, color = colors.textFaint)
            Text(
                "A PDR clip recorded with the track's beacon arrives with exact lap times and needs " +
                    "nothing from you. A GoPro clip, or a PDR one without beacons, has GPS but no lap " +
                    "markers — tap where the start/finish line is and every pass across it is timed.",
                style = type.xs,
                color = colors.textMuted,
                modifier = Modifier.padding(top = 6.dp),
            )
            Text(
                ".vbo and other logger files stay on the web app, where the screen is bigger and the " +
                    "SD card is already in the laptop.",
                style = type.xs,
                color = colors.textFaint,
                modifier = Modifier.padding(top = 6.dp),
            )
        }
    }
}

/**
 * Picking and parsing, and the states in between.
 *
 * The parse runs on the IO dispatcher in one job and checks for cancellation
 * between reads — a driver who picked the wrong 4 GB clip must be able to back
 * out. Held by `rememberScreenModel`, so a rotation mid-parse neither restarts
 * nor loses it.
 */
class ImportModel(
    private val scope: CoroutineScope,
    private val resolver: ContentResolver,
) {
    /** Non-null once a selection has been parsed; the review takes it from here. */
    var clips by mutableStateOf<List<ImportedClip>?>(null)
        private set

    var isParsing by mutableStateOf(false)
        private set

    var progress by mutableStateOf("")
        private set

    var failure by mutableStateOf<String?>(null)
        private set

    private var job: Job? = null

    fun parse(uris: List<Uri>) {
        if (uris.isEmpty()) return
        job?.cancel()
        failure = null
        isParsing = true
        progress = "Reading telemetry from ${uris.size} file${if (uris.size == 1) "" else "s"}…"
        job = scope.launch {
            try {
                clips = TelemetryImporter.parse(resolver, uris)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                failure = e.message ?: "Couldn't read the selected videos."
            } finally {
                isParsing = false
            }
        }
    }

    fun cancel() {
        job?.cancel()
        job = null
        isParsing = false
    }

    /** The review has taken the clips; a re-composition must not hand them over twice. */
    fun consumeClips() {
        clips = null
    }
}
